File: Data Structures & Algorithms/first-missing-positive/submission-5.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        for (int i{}; i < nums.size(); ++i) {
            if (nums[i] < 0) {
                nums[i] = 0;
            }
        }
        for (int i{}; i < nums.size(); ++i) {
            int num{nums[i]};
            if (num < 0)
                num *= -1;
            if (0 < num && num <= nums.size()) {
                if (nums[num -1] > 0) {
                    nums[num - 1] *= -1;
                }
                else if(nums[num - 1] == 0) {
                    nums[num - 1] = -1;
                }
            }
        }
        for (int i{}; i < nums.size(); ++i) {
            if (nums[i] >= 0){
                return i + 1;
            }
        }
        return nums.size() + 1;
    }
};

File: Data Structures & Algorithms/hand-of-straights/submission-5.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        if (hand.length % groupSize != 0) {
            return false;
        }
        HashMap<Integer, Integer> count = new HashMap<>();
        for (int num : hand) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }
        for (int num: hand) {
            if (count.get(num) != 0) {
                int start = num;
                while (count.getOrDefault(start - 1, 0) != 0) {
                    start = start - 1;
                }
                for (int i = 0; i < groupSize; ++i) {
                    if (count.getOrDefault(i + start, 0) == 0) {
                        return false;
                    }
                    count.put(i + start, count.get(i + start) - 1);
                }
            }
        }
        return true;
    }
}
